### PR TITLE
fix: N' attribute and subscript array contents

### PR DESCRIPTION
**Pull request by jirimosinger**
- Fix for [N' attribute issue](https://github.com/eclipse/che-che4z-lsp-for-hlasm/issues/231)
- Subscript array contents are now visible without the need to expand the variable
- Added debugger tests for special behavior of some system variables
- Added general tests for some system variables

### CODE REVIEW

File: parser_library/src/debugging/set_symbol_variable.cpp [hunk 1]
```diff
@@ -68,21 +56,54 @@ std::vector<variable_ptr> set_symbol_variable::values() const
...
 }
 
 size_t set_symbol_variable::size() const { return set_symbol_.size(); }
 
+std::string set_symbol_variable::get_string_value(const std::optional<int>& index) const
+{
+    std::string string_value;
+
+    if (!index && !is_scalar())
+        string_value = get_string_array_value();
```
Comment (slavek-kucera) [inline parser_library/src/debugging/set_symbol_variable.cpp:72]:
```suggestion
        return get_string_array_value();
```

File: parser_library/src/debugging/set_symbol_variable.h [hunk 2]
```diff
@@ -42,14 +42,15 @@ class set_symbol_variable : public variable
 
 private:
     template<typename T>
-    T get_value() const;
+    T get_value(const std::optional<int>& index) const;
```
Comment (slavek-kucera) [inline parser_library/src/debugging/set_symbol_variable.h:45]: given the size, these could be by-value... but I don't mind.

File: parser_library/test/debugging/debugger_test.cpp [hunk 3]
```diff
@@ -170,7 +162,7 @@ class test_var_value
 private:
     std::unordered_map<std::string, std::shared_ptr<test_var_value>> children_;
     std::optional<std::string> data_;
-    bool ignore_;
+    bool ignore_ = false;
```
Comment (slavek-kucera) [inline parser_library/test/debugging/debugger_test.cpp:165]: 👍

File: parser_library/test/debugging/debugger_test.cpp [hunk 4]
```diff
@@ -253,6 +262,46 @@ bool check_step(
...
+
+void step_into(debugger& d,
+    debug_event_consumer_s_mock& m,
+    std::vector<debugging::stack_frame>& exp_stack_frames,
+    size_t line,
+    std::string name,
+    debugging::source source)
+{
+    exp_stack_frames.insert(
+        exp_stack_frames.begin(), debugging::stack_frame(line, line, exp_stack_frames[0].id + 1, name, source));
```
Comment (slavek-kucera) [inline parser_library/test/debugging/debugger_test.cpp:290]: should this handle the empty case as well, just for consistency? or at least assert non-emptiness?

File: clients/vscode-hlasmplugin/CHANGELOG.md [hunk 5]
```diff
@@ -14,6 +14,8 @@
 - AINSERT can now do immediate variable evaluatation correctly
 - Highlighting of single character strings (which could represent data attributes - e.g. `I'`, `L'`, `S'` and others) is fixed
 - Document navigation does not work correctly in Untitled documents
+- N' attribute now returns 1 for empty subscript arrays ("()") which are provided to a macro from a caller's 'macro's symbolic parameter
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply clients/vscode-hlasmplugin/CHANGELOG.md:17]: Description simplified

File: parser_library/src/context/macro_param_data.cpp [hunk 6]
```diff
@@ -68,8 +68,12 @@ size_t macro_param_data_composite::size() const { return data_.size(); }
...
 
 macro_param_data_composite::macro_param_data_composite(std::vector<macro_data_ptr> value)
-    : macro_param_data_component(value.size())
-    , data_(move(value))
+    : macro_param_data_component(!value.empty() ? value.size() : 1)
+    , data_(!value.empty() ? std::move(value) : std::move([] {
+        std::vector<context::macro_data_ptr> vec;
+        vec.push_back(std::make_unique<context::macro_param_data_dummy>());
+        return vec;
+    }()))
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/context/macro_param_data.cpp:76]: 👍

File: parser_library/src/debugging/macro_param_variable.cpp [hunk 7]
```diff
@@ -25,11 +25,15 @@ macro_param_variable::macro_param_variable(const context::macro_param_base& para
...
+    {
         name_ = std::to_string(index_.back());
+        value_ = macro_param_.get_value(index_);
+    }
     else
+    {
         name_ = "&" + *macro_param_.id;
-
-    value_ = macro_param_.get_value(index_);
+        value_ = (&macro_param_)->macro_param_base::get_value();
```
[the remark this answers is not in the document]
Comment (jirimosinger) [reply parser_library/src/debugging/macro_param_variable.cpp:35]: 👍

File: parser_library/src/debugging/set_symbol_variable.cpp (hunk 1 above)
Comment (jirimosinger) [reply parser_library/src/debugging/set_symbol_variable.cpp:72]: 👍

File: parser_library/src/debugging/set_symbol_variable.h (hunk 2 above)
Comment (jirimosinger) [reply parser_library/src/debugging/set_symbol_variable.h:45]: I don't mind changing it either.

File: parser_library/test/debugging/debugger_test.cpp (hunk 4 above)
Comment (jirimosinger) [reply parser_library/test/debugging/debugger_test.cpp:290]: It sure can.

**Review by slavek-kucera — APPROVED** [review]
we'll need to wait for the release, before we merge.

**Pull request merged**